core/embed/trezorhal/unix/touch/touch.c

```diff
@@ -31,17 +31,158 @@
 
-static bool _touch_detected = false;
 static int _touch_x = 0;
 static int _touch_y = 0;
 
+// distance from the edge where arrow button swipe starts [px]
+static const int _btn_swipe_begin = 20;
+// length of the arrow button swipe [px]
+static const int _btn_swipe_length = 60;
+
+// A state machine to handle both mouse inputs (simulating touch) and arrow
+// buttons (to simulate scroll movements). The variable `input_state` is used to
+// ensure that arrow keys are not processed when mouse input is in progress and
+// that mouse actions are not processed while arrow button swipe is not
+// finished.
+typedef enum {
+  IDLE,
+  MOUSE_ACTIVE,
+  MOUSE_OUTSIDE,
+  BUTTON_SWIPE_INITIATED,
+  BUTTON_SWIPE_COMPLETED
+} touch_input_state_t;
+
+static touch_input_state_t input_state = IDLE;
+
 static bool is_inside_display(int x, int y) {
-  return (x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
-          x - sdl_touch_offset_x < sdl_display_res_x &&
-          y - sdl_touch_offset_y < sdl_display_res_y);
+  return x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
+         x - sdl_touch_offset_x < sdl_display_res_x &&
+         y - sdl_touch_offset_y < sdl_display_res_y;
+}
+
+static void handle_mouse_events(SDL_Event event, int* ev_type, int* ev_x,
+                                int* ev_y) {
+  bool inside_display = is_inside_display(event.button.x, event.button.y);
+  switch (event.type) {
+    case SDL_MOUSEBUTTONDOWN:
+      if (inside_display) {
+        *ev_x = event.button.x - sdl_touch_offset_x;
+        *ev_y = event.button.y - sdl_touch_offset_y;
+        *ev_type = TOUCH_START;
+        input_state = MOUSE_ACTIVE;
+      }
+      break;
+
+    case SDL_MOUSEBUTTONUP:
+      if (input_state != IDLE) {
+        *ev_x = inside_display ? event.button.x - sdl_touch_offset_x : _touch_x;
+        *ev_y = inside_display ? event.button.y - sdl_touch_offset_y : _touch_y;
+        *ev_type = TOUCH_END;
+        input_state = IDLE;
+      }
+      break;
+
+    case SDL_MOUSEMOTION:
+      if (input_state != IDLE) {
+        if (inside_display) {
+          *ev_x = event.motion.x - sdl_touch_offset_x;
+          *ev_y = event.motion.y - sdl_touch_offset_y;
+          // simulate TOUCH_START if pressed in mouse returned on visible area
+          *ev_type = (input_state == MOUSE_OUTSIDE) ? TOUCH_START : TOUCH_MOVE;
+          input_state = MOUSE_ACTIVE;
+        } else {
+          if (input_state == MOUSE_ACTIVE) {
+            // use last valid coordinates and simulate TOUCH_END
+            *ev_x = _touch_x;
+            *ev_y = _touch_y;
+            *ev_type = TOUCH_END;
+          }
+          input_state = MOUSE_OUTSIDE;
+        }
+      }
+      break;
+  }
+}
+
+static void handle_button_events(SDL_Event event, int* ev_type, int* ev_x,
+                                 int* ev_y) {
+  // Handle arrow buttons to trigger a scroll movement by set length in the
+  // direction of the button
+  switch (event.type) {
```

Comment:
Looks good; just consider avoiding nesting switch statements, for me it's not as clear as using if-switch structures.

```c
if (event.type == SDL_KEYDOWN && !event.key.repeat) {
  switch (..) {
      ...
  }
 } else if (event.type == SDL_KEYUP && input_state == BUTTON_SWIPE_INITIATED) {
  switch () {
     ..
  }
}
```

Reply:
refactored
https://github.com/trezor/trezor-firmware/pull/3619/commits/bedeb8a1be33d6fbc7c7ecb04adbfcab8514bb98